drivers/flash/nrf_qspi_nor.c

```diff
@@ -499,35 +528,82 @@ static int qspi_nor_read(struct device *dev, off_t addr, void *dest,
 		LOG_ERR("read error: address or size "
 			"exceeds expected values."
 			"Addr: 0x%lx size %zu", (long)addr, size);
+#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
+		ret = -EINVAL;
+		goto out;
+#else
 		return -EINVAL;
+#endif
 	}
 
 	qspi_lock(dev);
 
-	int ret = nrfx_qspi_read(dest, size, addr);
+	ret = nrfx_qspi_read(dest, size, addr);
 
 	qspi_wait_for_completion(dev);
+#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
+	if (rx_ram_buf != NULL) {
+		memcpy(tmp_dest, dest, var_size);
+		k_free(rx_ram_buf);
+	}
+#endif
 	return qspi_get_zephyr_ret_code(ret);
+#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
+out:
+	k_free(rx_ram_buf);
+	return ret;
+#endif
 }
 
 static int qspi_nor_write(struct device *dev, off_t addr, const void *src,
 			  size_t size)
 {
+	int ret;
+
 	if (!src) {
 		return -EINVAL;
 	}
 
+#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
+	tx_ram_buf = NULL;
+	if (((uint32_t)src < CONFIG_SRAM_BASE_ADDRESS)
+	   || (size < sizeof(uint32_t))) {
+		size_t var_size = size;
+
+		if (size < sizeof(uint32_t)) {
+			size = 4;
+		}
+		tx_ram_buf = k_malloc(size);
+		if (tx_ram_buf == NULL) {
+			ret = -ENOMEM;
+		}
+		qspi_nor_read(dev, addr, tx_ram_buf, size);
+		memcpy(tx_ram_buf, src, var_size);
+		src = tx_ram_buf;
+	}
+#endif
+
 	/* write size must be multiple of 4 bytes */
 	if (size % sizeof(uint32_t) ||
 	    !(size > 0)) {
+#if (CONFIG_HEAP_MEM_POOL_SIZE > 0)
+		ret = -EINVAL;
```

Comment:
See if you can do `goto out` in both these cases, then you can check the `if CONFIG_HEA_MEM_POOL_SIZE` inside the goto. This way the code flow won't be that different in the two different cases.

Reply:
This applies to this whole patch.

Reply:
Forgot to push the updated version to upstream fork :) Fixed.